### PR TITLE
[misc] feat: prototype deprecate DataProto and replace with Tensordict: part 2

### PR DESCRIPTION
**Pull request by houminz**
### What does this PR do?

This PR continues the work started in PR #2733, it adds support for variable sequence lengths in MultiTurnSFTDataset by introducing a `no_padding` option for the pad_mode. When this mode is active, sequences are not padded to a fixed length. 
- Implement no-padding mode for FSDP engine using nested tensors in sft trainer
- Add test for no-padding mode both enable/disable use_remove_padding
- Fix FSDP2 gradnorm issue

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 1]
```diff
@@ -781,11 +787,55 @@ def prepare_model_inputs(self, micro_batch: TensorDict):
             }
 
         else:
-            model_inputs = {
-                "input_ids": input_ids,
-                "attention_mask": attention_mask,
-                "position_ids": position_ids,
-            }
+            if pad_mode == DatasetPadMode.NO_PADDING:
```
Comment (vermouth1992) [inline verl/workers/engine/fsdp/transformer_impl.py:795]: There is actually no need to tackle no padding mode here. We should handle no padding mode if the model definition doesn't support rmpad inputs in L825 when use_remove_padding is False. SImply right pad to the longest sequence in the micro batch and perform rmpad to the output looks good

File: tests/test_protocol_v2_on_cpu.py [hunk 2]
```diff
@@ -63,29 +63,43 @@ def test_union_tensor_dict():
         tu.union_tensor_dict(data1, data2)
 
 
-def test_tensor_dict_constructor():
```
Comment (vermouth1992) [inline tests/test_protocol_v2_on_cpu.py:66]: Please do not remove tests

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 3]
```diff
@@ -426,3 +427,43 @@ def __getitem__(self, item):
...
+
+        Args:
+            batch: A list of dictionary samples from the dataset.
+
+        Returns:
+            A dictionary representing the batched data, with variable-length
+            sequences converted to NestedTensors.
+        """
+        # These keys correspond to tensors that will be nested.
+        tensor_keys = ["input_ids", "loss_mask", "position_ids"]
```
Comment (vermouth1992) [inline verl/utils/dataset/multiturn_sft_dataset.py]: Shall we introduce a general collator for all the cases that include both nested tensor and regular tensor, which is key agonistic?

File: verl/utils/seqlen_balancing.py [hunk 4]
```diff
@@ -273,11 +274,17 @@ def rearrange_micro_batches(
         List[List[int]]: index lists mapping each micro-batch back to original positions.
     """
     # this is per local micro_bsz
-    max_seq_len = batch["attention_mask"].shape[-1]
+    input_ids = batch["input_ids"]
```
Comment (vermouth1992) [inline verl/utils/seqlen_balancing.py:277]: This is a little bit ugly. Can we directly pass in `seq_len_effective` as an argument?

Comment (vermouth1992) [inline verl/utils/seqlen_balancing.py:277]: We may want to modify this function by directly passing `seq_len_effective`

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 5]
```diff
@@ -781,11 +787,55 @@ def prepare_model_inputs(self, micro_batch: TensorDict):
...
+                max_prompt_length = tu.get_non_tensor_data(data=micro_batch, key="max_prompt_length", default=1024)
```
Comment (vermouth1992) [inline verl/workers/engine/fsdp/transformer_impl.py]: We can simply use right truncation here using the max length

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 6]
```diff
@@ -426,3 +427,43 @@ def __getitem__(self, item):
                 "responses": response_ids,
                 "response_mask": response_loss_mask,
             }
+        elif self.pad_mode == DatasetPadMode.NO_PADDING:
+            # create position IDs
+            position_ids = torch.arange(len(input_ids), dtype=torch.long)
+            # return nested tensor with out padding
+            return {
+                "input_ids": input_ids,
```
Comment (vermouth1992) [inline verl/utils/dataset/multiturn_sft_dataset.py:439]: We still need to truncate using max_length

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 7]
```diff
@@ -707,38 +709,44 @@ def prepare_model_inputs(self, micro_batch: TensorDict):
...
-            if position_ids.dim() == 3:
-                position_ids_rmpad = (
-                    index_first_axis(rearrange(position_ids, "c b s ... -> (b s) c ..."), indices)
-                    .transpose(0, 1)
-                    .unsqueeze(1)
-                )  # (3, bsz, seqlen) -> (3, 1, bsz * seqlen)
+            if pad_mode == DatasetPadMode.NO_PADDING:
+                input_ids_rmpad = input_ids.values().unsqueeze(0)  # (1, total_nnz)
+                position_ids_rmpad = position_ids.values().unsqueeze(0)  # (1, total_nnz)
             else:
```
Comment (vermouth1992) [inline verl/workers/engine/fsdp/transformer_impl.py]: Add elif DatasetPadMode.LEFT_RIGHT

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 8]
```diff
@@ -707,38 +709,44 @@ def prepare_model_inputs(self, micro_batch: TensorDict):
...
+                if position_ids.dim() == 3:
+                    position_ids_rmpad = (
+                        index_first_axis(rearrange(position_ids, "c b s ... -> (b s) c ..."), indices)
+                        .transpose(0, 1)
+                        .unsqueeze(1)
+                    )  # (3, bsz, seqlen) -> (3, 1, bsz * seqlen)
+                else:
+                    position_ids_rmpad = index_first_axis(
+                        rearrange(position_ids.unsqueeze(-1), "b s ... -> (b s) ..."), indices
+                    ).transpose(0, 1)
```
Comment (vermouth1992) [inline verl/workers/engine/fsdp/transformer_impl.py:745]: Add else: raise NotImplementError

File: tests/test_protocol_v2_on_cpu.py (hunk 2 above)
Comment (houminz) [reply tests/test_protocol_v2_on_cpu.py:66]: updated

File: verl/utils/dataset/multiturn_sft_dataset.py (hunk 6 above)
Comment (houminz) [reply verl/utils/dataset/multiturn_sft_dataset.py:439]: done

File: verl/utils/seqlen_balancing.py (hunk 4 above)
Comment (houminz) [reply verl/utils/seqlen_balancing.py:277]: would update this function when we modify the rl trainer

File: verl/workers/engine/fsdp/transformer_impl.py (hunk 8 above)
Comment (houminz) [reply verl/workers/engine/fsdp/transformer_impl.py:745]: done

File: verl/workers/engine/fsdp/transformer_impl.py (hunk 5 above)
Comment (houminz) [reply verl/workers/engine/fsdp/transformer_impl.py]: updated

**Comment by houminz** [comment]
Hi @vermouth1992 , Thank you again for the thorough code review. I truly appreciate your time and feedback.

I have now revised the code based on your comments. Please let me know if the updates look good when you have a moment for a second review.

File: verl/utils/dataset/dataset_utils.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2025 ModelBest Inc. and/or its affiliates
```
Comment (vermouth1992) [inline verl/utils/dataset/dataset_utils.py]: The license should be Bytedance

File: verl/utils/dataset/dataset_utils.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        if self.pad_mode == DatasetPadMode.NO_PADDING:
+            return self.collate_variable_batch(batch)
+        elif self.pad_mode in [DatasetPadMode.RIGHT, DatasetPadMode.LEFT_RIGHT]:
+            from torch.utils.data import default_collate
+
+            return default_collate(batch)
+        else:
+            raise NotImplementedError(f"pad_mode {self.pad_mode} not implemented")
+
+    def collate_variable_batch(self, batch: list[dict[str, any]]) -> dict[str, any]:
```
Comment (vermouth1992) [inline verl/utils/dataset/dataset_utils.py:49]: Can this collator handle mixed nested tensor and regular tensor?
